File: web/src/api/endpoints/error-log.ts
import { useInfiniteQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { logger } from '@/lib/logger';

/**
 * 错误日志条目（前后端崩溃/错误，与转发日志 relay_logs 分离）。
 */
export interface ErrorLog {
    id: number;
    time: number;              // 时间戳（秒）
    source: 'backend' | 'frontend'; // 来源
    level: string;             // panic | error | unhandledrejection | uncaught
    message: string;           // 错误信息
    stack: string;             // 完整堆栈
    request_method?: string;   // 后端请求方法
    request_path?: string;     // 后端请求路径
    client_ip?: string;
    user_agent?: string;
    page_url?: string;         // 前端页面 URL
    route_id?: string;         // 前端路由/模块 id
    version?: string;          // 上报方版本
}

export interface ErrorLogFilter {
    source?: 'backend' | 'frontend';
    level?: string;
    start_time?: number;
    end_time?: number;
}

export const DEFAULT_ERROR_LOG_PAGE_SIZE = 20;

const errorLogsQueryKey = (pageSize: number, filter: ErrorLogFilter) => ['error-logs', 'infinite', pageSize, filter] as const;

/**
 * 错误日志分页查询 Hook（滚动加载）。
 */
export function useErrorLogs(options: { pageSize?: number; filter?: ErrorLogFilter } = {}) {
    const { pageSize = DEFAULT_ERROR_LOG_PAGE_SIZE, filter = {} } = options;
    return useInfiniteQuery({
        queryKey: errorLogsQueryKey(pageSize, filter),
        initialPageParam: 1,
        queryFn: async ({ pageParam }) => {
            const params = new URLSearchParams();
            params.set('page', String(pageParam));
            params.set('page_size', String(pageSize));
            if (filter.source) params.set('source', filter.source);
            if (filter.level) params.set('level', filter.level);
            if (filter.start_time != null) params.set('start_time', String(filter.start_time));
            if (filter.end_time != null) params.set('end_time', String(filter.end_time));
            const result = await apiClient.get<ErrorLog[] | null>(`/api/v1/error-log/list?${params.toString()}`);
            return result ?? [];
        },
        getNextPageParam: (lastPage, allPages) => {
            if (!lastPage || lastPage.length < pageSize) return undefined;
            return allPages.length + 1;
        },
        staleTime: 30_000,
    });
}

/**
 * 清空错误日志。
 */
export function useClearErrorLogs() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async () => {
            return apiClient.delete<null>('/api/v1/error-log/clear');
        },
        onSuccess: () => {
            logger.log('错误日志清空成功');
            queryClient.invalidateQueries({ queryKey: ['error-logs'] });
        },
        onError: (error) => {
            logger.error('错误日志清空失败:', error);
        },
    });
}

File: internal/model/error_log.go
package model

// ErrorLog 记录前后端崩溃/错误日志（排障用）。
//
// 来源（Source）：
//   - backend：后端 Go panic（Gin CustomRecovery 捕获）或系统级错误
//   - frontend：管理 UI 捕获的 JS 错误（window.onerror / unhandledrejection /
//     React ErrorBoundary）
//
// Level 区分严重程度：panic（崩溃）、error、unhandledrejection、uncaught。
// Message 为错误简述，Stack 为完整堆栈；请求/页面信息辅助定位触发场景。
type ErrorLog struct {
	ID     int64  `json:"id" gorm:"primaryKey;autoIncrement:false"` // Snowflake ID
	Time   int64  `json:"time" gorm:"column:time;index"`            // 时间戳（秒）
	Source string `json:"source" gorm:"column:source;size:16"`      // backend | frontend
	Level  string `json:"level" gorm:"column:level;size:32"`        // panic | error | unhandledrejection | uncaught
	// Message 错误信息（panic 值 / JS 错误信息）
	Message string `json:"message" gorm:"column:message;type:text"`
	// Stack 完整堆栈（后端 debug.Stack() / 前端 error.stack）
	Stack string `json:"stack" gorm:"column:stack;type:text"`
	// 后端请求信息（panic 发生时）
	RequestMethod string `json:"request_method" gorm:"column:request_method;size:16"`
	RequestPath   string `json:"request_path" gorm:"column:request_path;size:512"`
	ClientIP      string `json:"client_ip" gorm:"column:client_ip;size:64"`
	UserAgent     string `json:"user_agent" gorm:"column:user_agent;size:512"`
	// 前端页面信息（错误发生时）
	PageURL string `json:"page_url" gorm:"column:page_url;size:1024"`
	RouteID string `json:"route_id" gorm:"column:route_id;size:128"`
	// 上报方版本（后端 conf.Version / 前端 build 版本）
	Version string `json:"version" gorm:"column:version;size:64"`
}

// TableName 显式指定表名（与 GORM 默认一致，便于阅读）。
func (ErrorLog) TableName() string { return "error_logs" }
